File: android-vinetrack/app/src/main/java/com/rork/vinetrack/data/spray/SprayTarget.kt
package com.rork.vinetrack.data.spray

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * What a spray is aimed at — the pest, disease or agronomic purpose. The Kotlin
 * twin of the Swift `SprayTarget`.
 *
 * Stored as a STABLE INTERNAL IDENTIFIER, never as display text, so the label can
 * be reworded (or localised) without orphaning historical records and without
 * breaking the future Resistance Check, which will match on these identifiers
 * rather than on strings a user typed.
 *
 * Multiple targets per spray are supported by design: a single tank commonly
 * addresses powdery and downy in one pass.
 */
@Serializable
enum class SprayTarget(val raw: String, val label: String) {
    @SerialName("powdery_mildew")
    POWDERY_MILDEW("powdery_mildew", "Powdery Mildew"),

    @SerialName("downy_mildew")
    DOWNY_MILDEW("downy_mildew", "Downy Mildew"),

    @SerialName("botrytis")
    BOTRYTIS("botrytis", "Botrytis"),

    @SerialName("weeds")
    WEEDS("weeds", "Weeds"),

    @SerialName("nutrition_biostimulant")
    NUTRITION_BIOSTIMULANT("nutrition_biostimulant", "Nutrition / Biostimulant"),

    @SerialName("other")
    OTHER("other", "Other"),
    ;

    /**
     * True for targets a fungicide resistance strategy applies to.
     *
     * This is the hook the Resistance Check will read. It deliberately lives on
     * the target rather than in the UI so the future rules engine and this flow
     * cannot disagree about which targets are resistance-relevant.
     */
    val isFungicideResistanceRelevant: Boolean
        get() = when (this) {
            POWDERY_MILDEW, DOWNY_MILDEW, BOTRYTIS -> true
            WEEDS, NUTRITION_BIOSTIMULANT, OTHER -> false
        }

    companion object {
        /**
         * Tolerant decode from stored text, so an unknown or legacy value degrades
         * to null instead of failing the record.
         */
        fun from(raw: String?): SprayTarget? {
            val value = raw?.trim()?.lowercase()?.takeIf { it.isNotEmpty() } ?: return null
            entries.firstOrNull { it.raw == value }?.let { return it }
            return when (value) {
                "powdery", "pm", "powdery mildew" -> POWDERY_MILDEW
                "downy", "dm", "downy mildew" -> DOWNY_MILDEW
                "nutrition", "biostimulant", "foliar nutrition" -> NUTRITION_BIOSTIMULANT
                "weed", "herbicide" -> WEEDS
                else -> null
            }
        }

        /**
         * Stable display order for pickers on both platforms, so iOS and Android
         * present the same decisions in the same sequence.
         */
        val presentationOrder: List<SprayTarget> = listOf(
            POWDERY_MILDEW, DOWNY_MILDEW, BOTRYTIS, WEEDS, NUTRITION_BIOSTIMULANT, OTHER,
        )
    }
}

/**
 * Where the spray head is aimed for a foliar application.
 *
 * Extensible on purpose: other Australian and New Zealand terminology will be
 * added without changing the persisted shape, because this is stored as a stable
 * identifier too.
 */
@Serializable
enum class SprayHeadTarget(val raw: String, val label: String, val detail: String) {
    @SerialName("full_canopy")
    FULL_CANOPY("full_canopy", "Full Canopy", "Whole canopy, top to bottom"),

    @SerialName("bunch_line")
    BUNCH_LINE("bunch_line", "Bunch Line", "Fruit zone only"),

    @SerialName("leaf_zone")
    LEAF_ZONE("leaf_zone", "Leaf Zone", "Foliage above the fruit zone"),
    ;

    companion object {
        fun from(raw: String?): SprayHeadTarget? {
            val value = raw?.trim()?.lowercase()?.takeIf { it.isNotEmpty() } ?: return null
            entries.firstOrNull { it.raw == value }?.let { return it }
            return when (value) {
                "full canopy", "canopy", "whole canopy" -> FULL_CANOPY
                "bunch line", "bunchline", "fruit zone" -> BUNCH_LINE
                "leaf zone", "leafzone", "foliage" -> LEAF_ZONE
                else -> null
            }
        }
    }
}

/**
 * The operator-facing application method. Raw values match the iOS
 * `OperationType` enum and the existing `sprayOperationTypes` strings, so the
 * stored `operation_type` column is unchanged.
 */
enum class SprayOperationType(val raw: String) {
    FOLIAR_SPRAY("Foliar Spray"),
    BANDED_SPRAY("Banded Spray"),
    SPREADER("Spreader"),
    ;

    companion object {
        fun from(raw: String?): SprayOperationType? {
            val value = raw?.trim() ?: return null
            return entries.firstOrNull { it.raw.equals(value, ignoreCase = true) }
        }
    }
}

File: ios/VineTrack/App/SprayEngine/SprayTarget.swift
import Foundation

/// What a spray is aimed at — the pest, disease or agronomic purpose.
///
/// Stored as a STABLE INTERNAL IDENTIFIER, never as display text, so the label
/// can be reworded (or localised) without orphaning historical records and
/// without breaking the future Resistance Check, which will match on these
/// identifiers rather than on strings a user typed.
///
/// Multiple targets per spray are supported by design: a single tank commonly
/// addresses powdery and downy in one pass.
nonisolated enum SprayTarget: String, Sendable, Codable, CaseIterable, Hashable, Identifiable {
    case powderyMildew = "powdery_mildew"
    case downyMildew = "downy_mildew"
    case botrytis = "botrytis"
    case weeds = "weeds"
    case nutritionBiostimulant = "nutrition_biostimulant"
    case other = "other"

    nonisolated var id: String { rawValue }

    var label: String {
        switch self {
        case .powderyMildew: return "Powdery Mildew"
        case .downyMildew: return "Downy Mildew"
        case .botrytis: return "Botrytis"
        case .weeds: return "Weeds"
        case .nutritionBiostimulant: return "Nutrition / Biostimulant"
        case .other: return "Other"
        }
    }

    var iconName: String {
        switch self {
        case .powderyMildew: return "circle.dotted"
        case .downyMildew: return "drop.triangle"
        case .botrytis: return "allergens"
        case .weeds: return "leaf"
        case .nutritionBiostimulant: return "sparkles"
        case .other: return "questionmark.circle"
        }
    }

    /// True for targets a fungicide resistance strategy applies to.
    ///
    /// This is the hook the Resistance Check will read. It deliberately lives on
    /// the target rather than in the UI so the future rules engine and this flow
    /// cannot disagree about which targets are resistance-relevant.
    var isFungicideResistanceRelevant: Bool {
        switch self {
        case .powderyMildew, .downyMildew, .botrytis: return true
        case .weeds, .nutritionBiostimulant, .other: return false
        }
    }

    /// Tolerant decode from stored text, so an unknown or legacy value degrades
    /// to `nil` instead of failing the record.
    static func from(_ raw: String?) -> SprayTarget? {
        guard let raw = raw?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased(),
              !raw.isEmpty else { return nil }
        if let exact = SprayTarget(rawValue: raw) { return exact }
        switch raw {
        case "powdery", "pm", "powdery mildew": return .powderyMildew
        case "downy", "dm", "downy mildew": return .downyMildew
        case "nutrition", "biostimulant", "foliar nutrition": return .nutritionBiostimulant
        case "weed", "herbicide": return .weeds
        default: return nil
        }
    }

    /// Stable display order for pickers on both platforms, so iOS and Android
    /// present the same decisions in the same sequence.
    static let presentationOrder: [SprayTarget] = [
        .powderyMildew, .downyMildew, .botrytis, .weeds, .nutritionBiostimulant, .other
    ]
}

/// Where the spray head is aimed for a foliar application.
///
/// Extensible on purpose: other Australian and New Zealand terminology will be
/// added without changing the persisted shape, because this is stored as a
/// stable identifier too.
nonisolated enum SprayHeadTarget: String, Sendable, Codable, CaseIterable, Hashable, Identifiable {
    case fullCanopy = "full_canopy"
    case bunchLine = "bunch_line"
    case leafZone = "leaf_zone"

    nonisolated var id: String { rawValue }

    var label: String {
        switch self {
        case .fullCanopy: return "Full Canopy"
        case .bunchLine: return "Bunch Line"
        case .leafZone: return "Leaf Zone"
        }
    }

    var detail: String {
        switch self {
        case .fullCanopy: return "Whole canopy, top to bottom"
        case .bunchLine: return "Fruit zone only"
        case .leafZone: return "Foliage above the fruit zone"
        }
    }

    static func from(_ raw: String?) -> SprayHeadTarget? {
        guard let raw = raw?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased(),
              !raw.isEmpty else { return nil }
        if let exact = SprayHeadTarget(rawValue: raw) { return exact }
        switch raw {
        case "full canopy", "canopy", "whole canopy": return .fullCanopy
        case "bunch line", "bunchline", "fruit zone": return .bunchLine
        case "leaf zone", "leafzone", "foliage": return .leafZone
        default: return nil
        }
    }
}
